### ISSUE TITLE
report.render hang most of the time

### ISSUE THREAD

**Issue by kahwooi** [jsreport/jsreport#39]
I am using jsreport with jade and express, but most of time it hang during report.render stage. I have no clue why?

```
         var html = jade.renderFile(path.resolve(__dirname + '/..') + '/views/report/report.jade',  JSON.parse(data));

                report.render({
                    template:{
                        content: html,
                        phantom: {
                            footer: "<div style='text-align:center'>{#pageNum}/{#numPages}</div>"
                        }
                    }
                }).then(function (out) {
                    out.result.pipe(res);
                });
```

**Comment by pofider** [comment]
1.Do you also catch failed promise? 
```js
report.render({...})
  .then(function (out) {
    out.result.pipe(res);
  }).catch(function(e) {
    console.log(e);
  });
```

2.Is there a difference when running without attached debugger? Some IDEs are having problems with child processes, see https://github.com/jsreport/jsreport/issues/25

3.You can also add a console logger to see where it stucs
http://jsreport.net/learn/pdf-reports-in-nodejs
```js
jsreport.renderDefaults.logger.providerName = "console";
```

**Comment by kahwooi** [comment]
Yes, I think it is the problem 2. It happen when I Webstorm debugging mode.

**Comment by pofider** [comment]
Unfortunately now you have only two options. Run jsreport in a separate server and use [jsreport remote client](http://jsreport.net/learn/nodejs-client) for communication or start your app without debugger. 

I understand this can be very annoying. I am waiting already for a year for fix in the IDE but it doesn't seems to be coming. I will increase the priority for solving it in the jsreport https://github.com/jsreport/jsreport/issues/25

**Comment by kahwooi** [comment]
Is there any configuration to start jsreport without child processes during development?

**Comment by pofider** [comment]
Not yet, I should get to it in the next weeks.

**Comment by kahwooi** [comment]
I also found out that the first time report loading is slow, but subsequent report request is much faster. Any settings I can do to make it faster. Usually the first time request does not return any result, I have to refresh the page to get a result.

**Comment by pofider** [comment]
It takes like one second on my machine. How long does it take for you? 

jsreport crawls node_modules/jsreport folder to find applicable extensions, this can be slow with the slow disk. I can add configuration option to skip this if it turns out startup time is major problem.

To speed up the first rendering you can just render a blank report during your app startup. Would it be a solution for you?

**Comment by kahwooi** [comment]
I got this after sometime:

```
{ [Error: Error during rendering report: Timeout]
  weak: true,
  message: 'Error during rendering report: Timeout' }
```

**Comment by pofider** [comment]
The configuration files [prod|dev].config.json are used when you start jsreport using [bootstrapper](http://jsreport.net/learn/adapting-jsreport). Which one is optionally pre-created and used depends on the NODE_ENV node.js variable.

It is really strange to me your performance problems. The jsreportonline service is rendering thousands pdf reports per single core machine. There must be something wrong. Would you be able to separate and send me a minimal "not working" example?

To answer your question, you can set the array of extensions to apply in the [configuration file](https://github.com/jsreport/jsreport/blob/master/config.md). This will then shorten the the rendering pipeline.
```js
{
 ...
 extensions: ["html", "templates", "data", "phantom-pdf"]
}

**Comment by kahwooi** [comment]
The latest updates does not solve this problem for webstorm debugging mode.

**Comment by pofider** [comment]
It works for me on:
win 8.1
jsreport 0.2.10
webstorm 9.0.1
node 0.10.26

what is your env?

**Comment by kahwooi** [comment]
Mine is: 
 
 Windows 8.1
 jsreport 0.2.10
 webstorm 9.0.3
 node 0.12.0

**Comment by pofider** [comment]
I have checked it again with the same env as yours. It seems that it doesn't work with x64 node. Can you install x86 one and switch it in webstorm? Thanks.

**Comment by kahwooi** [comment]
I have tried to switch it to node x86, I had waited for more the 10 minutes, and it still stuck at:

```
Should be empty: []
``` 

It takes forever to start now.

**Comment by pofider** [comment]
Other people are complaining slow startup now as well. Unfortunately I am still not able to find a reason for this.

Isn't the size of your "data" folder significantly big? Because jsreport is loading everything from the data folder into memory at the startup, so this could be the problem.

**Comment by BoVut** [comment]
Mine was also slow, and my friend could not pull the project from the source control. He found that there were around 4500+ files in the App_Data (the folder I specified in the Global.asax when set up the jsreport). After I'm not sure what's inside, but I think I could possibly be the cause.

**Comment by pofider** [comment]
I think this issue was fixed in version 0.8 by changing the default rendering strategy. See this blog post for details http://jsreport.net/blog/performance-impacting-change-in-v08

I haven't see anything similar to this since then, it is the closed topic to me. Feel free to reopen if you have this problem on the new versions.

**Comment by MihaelIsaev** [comment]
As for me it works with debugger only if I set strategy to `in-process`

```
let jsreport = require('jsreport-core')({ tasks: { strategy: 'in-process' } })
```
